app/upgrades/v1-fork-btc-reorg-k/testnet/upgrade.go

```diff
@@ -0,0 +1,349 @@
+package testnet
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/babylonlabs-io/babylon/app/keepers"
+	"github.com/babylonlabs-io/babylon/app/upgrades"
+	bbn "github.com/babylonlabs-io/babylon/types"
+	bskeeper "github.com/babylonlabs-io/babylon/x/btcstaking/keeper"
+	bstypes "github.com/babylonlabs-io/babylon/x/btcstaking/types"
+	fkeeper "github.com/babylonlabs-io/babylon/x/finality/keeper"
+	ftypes "github.com/babylonlabs-io/babylon/x/finality/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"go.uber.org/zap"
+)
+
+const (
+	UpgradeName = "v1-btc-reorg-k"
+)
+
+var (
+	// MapUnbondStkTxHashRollback this should be the staking transaction
+	// hash hex of the unbonding transactions that got rolled back.
+	MapUnbondStkTxHashRollback = map[string]struct{}{
+		// example staking tx of an unbonding staking transaction
+		"a3d84be950961d03a72c04a0128cce000e3476f489e17273cbea71f971b47f61": {},
+	}
+)
+
+func CreateUpgrade() upgrades.Fork {
+	return upgrades.Fork{
+		UpgradeName: UpgradeName,
+		// TODO: fill with correct block height of fork,
+		UpgradeHeight:  12000,
+		BeginForkLogic: CreateForkLogic,
+	}
+}
+
+// CreateForkLogic executes the fork logic to handle BTC reorg large than k.
+func CreateForkLogic(context sdk.Context, keepers *keepers.AppKeepers) {
+	err := ForkHandler(context, keepers)
+	if err != nil {
+		panic(fmt.Errorf("failed to run the fork handler: %w", err))
+	}
+}
+
+// ForkHandler wraps the logic of the fork to return an error
+func ForkHandler(context sdk.Context, keepers *keepers.AppKeepers) error {
```

Comment:
should we include also the logic to rollback headers in the btclightclient mod?

Reply:
The rollback of BTC headers in the [btclightclient](https://github.com/babylonlabs-io/babylon/blob/dbcc7c4c72141240f2770754109437c735977401/x/btclightclient/keeper/keeper.go#L103) and [btccheckpoint](https://github.com/babylonlabs-io/babylon/blob/dbcc7c4c72141240f2770754109437c735977401/x/btccheckpoint/keeper/hooks.go#L26) are already handled on the spot, no need to have modifications in the fork handler

Correct me if I am wrong @KonradStaniec